R/utils.R

```diff
@@ -485,15 +472,122 @@ multi_choice_to_labels <- function(db_data, db_metadata) {
 
       # Replace values from db_data$(field_name) with label values from
       # parse_labels key
-      db_data[[field_name]] <- db_data[[field_name]] %>%
-        as.character() %>%
-        recode(!!!parse_labels_output) %>%
-        factor(levels = unique(parse_labels_output))
+
+      db_data[[field_name]] <- label_handler(
+        x = db_data[[field_name]],
+        labels = parse_labels_output,
+        ptype = db_data[[field_name]]
+      )
     }
   }
   db_data
 }
 
+#' @title
+#' Apply factor labels to a vector
+#'
+#' @param x a vector to label
+#' @param labels a named vector of labels in the format `c(value = label)`
+#' @param \dots unused
+#'
+#' @return
+#' factor
+#'
+#' @keywords internal
+apply_labs_factor <- function(x, labels, ...) {
+  as.character(x) %>%
+    recode(!!!labels) %>%
+    factor(levels = unique(labels))
+}
+
+#' @title
+#' Apply haven value labels to a vector
+#'
+#' @details
+#' Assumes a check_installed() has been run for `labelled`. Since `haven` preserves the
+#' underlying data values we need to make sure the data type of the value options in the metadata matches
+#' the data type of the values in the actual data. This function accepts a prototype, usually a column
+#' from db_data, and uses `force_cast()` to do a best-effort casting of the value options in the metadata
+#' to the same data type as `ptype`. The fallback is to convert `x` and the value labels to character.
+#'
+#' @param x a vector to label
+#' @param labels a named vector of labels in the format `c(value = label)`
+#' @param ptype vector to serve as prototype for label values
+#' @param \dots unused
```

Comment:
Ditto above

Reply:
In this case we actually don't need it since the parameters of `apply_labs_haven()` are a superset of `apply_labs_factor()`. I kept it around to mirror `apply_labs_factor` but we can remove